extensions/amp-auto-ads/0.1/firstimpression.io-network-config.js

```diff
@@ -0,0 +1,144 @@
+import {buildUrl} from '../../../ads/google/a4a/shared/url-builder';
+import {dict} from '../../../src/utils/object';
+
+/**
+ * @implements {./ad-network-config.AdNetworkConfigDef}
+ */
+export class FirstImpressionIoConfig {
+  /**
+   * @param {!Element} autoAmpAdsElement
+   */
+  constructor(autoAmpAdsElement) {
+    this.autoAmpAdsElement_ = autoAmpAdsElement;
+  }
+
+  /**
+   * @param {!Window} unused
+   * @override
+   */
+  isEnabled(unused) {
+    return true;
+  }
+
+  /**
+   * True if responsive is enabled for auto-ads
+   */
+  isResponsiveEnabled() {
+    return false;
+  }
+
+  /** @override */
+  getConfigUrl() {
+    let queryParams = {};
+    let previewId = 0;
+
+    const previewFlowRegex = /amp\/fi\/(\d+)\//;
+    const previewFlowParam = window.location.pathname.match(previewFlowRegex);
+    if (previewFlowParam != null && previewFlowParam.length == 2) {
+      previewId = previewFlowParam[1];
+    }
+
+    const docInfo = Services.documentInfoForDoc(this.autoAmpAdsElement_);
+
+    const fiReveal = this.getURLHashParameter_('fi_reveal');
+    const fiDemand = this.getURLHashParameter_('fi_demand');
+    const fiGeo = this.getURLHashParameter_('fi_geo');
+
+    const cdnHost =
+      this.getURLHashParameter_('fi_cdnhost') ||
+      (previewId ? window.location.host : 'cdn.firstimpression.io');
+    const cdnpath =
+      this.getURLHashParameter_('fi_cdnpath') ||
+      (previewId ? '/amp-preview.php' : '/delivery/amp.php');
+
+    const websiteId = this.autoAmpAdsElement_.getAttribute('data-website-id');
+    const targeting = this.autoAmpAdsElement_.getAttribute('data-targeting');
+
+    queryParams = {
+      'id': websiteId,
+      't': targeting,
+      'url': docInfo.canonicalUrl,
+      'w': window.screen.width,
+      'h': window.screen.height,
+    };
+
+    if (fiReveal) {
+      queryParams['fi_reveal'] = fiReveal;
+    }
+    if (fiDemand) {
+      queryParams['fi_demand'] = fiReveal;
+    }
+    if (fiGeo) {
+      queryParams['fi_geo'] = fiGeo;
+    }
+    if (previewId) {
+      queryParams['preview_id'] = previewId;
+    }
+
+    return buildUrl(
+      'https://' + cdnHost + cdnpath,
+      queryParams,
+      /* maxLength */ 4096
+    );
+  }
+
+  /** @override */
+  getAttributes() {
+    const attributes = dict({
+      'type': 'firstimpression',
+    });
+    return attributes;
+  }
+
+  /** @override */
+  getDefaultAdConstraints() {
+    const viewportHeight = Services.viewportForDoc(
+      this.autoAmpAdsElement_
+    ).getSize().height;
+    return {
+      initialMinSpacing: viewportHeight,
+      subsequentMinSpacing: [
+        {adCount: 3, spacing: viewportHeight * 2},
+        {adCount: 6, spacing: viewportHeight * 3},
+      ],
+      maxAdCount: 8,
+    };
+  }
+
+  /** @override */
+  getSizing() {
+    return {
+      height: 1,
+    };
+  }
+
+  /**
+   * @param {string} name fragment parameter name to retrieve
+   */
+  getURLHashParameter_(name) {
```

Comment:
also `parseUrlWithA` ? link above.

Reply:
for this one since `data.location` is not available you can use `Services.urlForDoc(this.element).parse(url).hash`

Reply:
even better is probably `this.win.location.hash`

Reply:
The context of this is ./ad-network-config.AdNetworkConfigDef, not amp element. Hence, the only available option that I see is using `this.element.ampdoc_.win.location.hash`. Will that be better then direct access to `window.location.hash`?

Reply:
yes you are totally right. Also I see the other auto ads impls using window directly, so I see why you also did. I think getting it from ampdoc is the right approach though (we probably should change the others as well :( ) Thanks!

Reply:
the autoads el passed in the constructor should have the right win so I think you can just store a ref there
`this.win = autoAdsElement.win`

Reply:
No, unfortunately the element passed to constructor does not contain `win` reference. The only reference to window that I can find in constructor fn parameter is `ampdoc_.win`

Reply:
thanks for bearing with me on this!